witness/golang/omniwitness/usbarmory/storage/slots/journal_test.go

```diff
@@ -0,0 +1,231 @@
+			}
+		})
+	}
+}
+
+func TestEntryUnmarshal(t *testing.T) {
+	for _, test := range []struct {
+		name    string
+		want    *entry
+		b       []byte
+		wantErr bool
+	}{
+		{
+			name: "golden",
+			want: &entry{
+				Magic:      magic0Hdr(),
+				Revision:   42,
+				DataLen:    uint64(len("hello")),
+				DataSHA256: sha256.Sum256([]byte("hello")),
+				Data:       []byte("hello"),
+			},
+			b: []byte{
+				'T', 'F', 'J', '0',
+				0x00, 0x00, 0x00, 0x2a,
+				0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x05,
+				0x2c, 0xf2, 0x4d, 0xba, 0x5f, 0xb0, 0xa3, 0x0e, 0x26, 0xe8, 0x3b, 0x2a, 0xc5, 0xb9, 0xe2, 0x9e, 0x1b, 0x16, 0x1e, 0x5c, 0x1f, 0xa7, 0x42, 0x5e, 0x73, 0x04, 0x33, 0x62, 0x93, 0x8b, 0x98, 0x24,
+				'h', 'e', 'l', 'l', 'o',
+			},
+		}, {
+			name: "bad magic",
+			b: []byte{
+				'N', 'o', 'P', 'e',
+				0x00, 0x00, 0x00, 0x2a,
+				0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x05,
+				0x2c, 0xf2, 0x4d, 0xba, 0x5f, 0xb0, 0xa3, 0x0e, 0x26, 0xe8, 0x3b, 0x2a, 0xc5, 0xb9, 0xe2, 0x9e, 0x1b, 0x16, 0x1e, 0x5c, 0x1f, 0xa7, 0x42, 0x5e, 0x73, 0x04, 0x33, 0x62, 0x93, 0x8b, 0x98, 0x24,
+				'h', 'e', 'l', 'l', 'o',
+			},
+			wantErr: true,
+		}, {
+			name: "bad hash",
+			b: []byte{
+				'N', 'o', 'P', 'e',
+				0x00, 0x00, 0x00, 0x2a,
+				0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x05,
+				0x2c, 0xf2, 0x4d, 0xba, 0x5f, 0xb0, 0xa3, 0x0e, 0x26, 0xe8, 0x3b, 0x2a, 0xc5, 0xb9, 0xe2, 0x9e, 0x1b, 0x16, 0x1e, 0x5c, 0x1f, 0xa7, 0x42, 0x5e, 0x73, 0x04, 0x33, 0x62, 0x93, 0x8b, 0x98, 0x24,
+				'W', 'R', 'O', 'N', 'G',
+			},
+			wantErr: true,
+		},
+	} {
+		t.Run(test.name, func(t *testing.T) {
+			got, err := unmarshalEntry(bytes.NewReader(test.b))
+			if gotErr := err != nil; gotErr != test.wantErr {
+				t.Fatalf("unmarshalEntry: %v, wantErr %t", err, test.wantErr)
+			}
+			if test.wantErr {
+				return
+			}
+
+			if diff := cmp.Diff(got, test.want); diff != "" {
+				t.Fatalf("Got diff: %v", diff)
+			}
+		})
+	}
+}
+
+func TestOpenJournal(t *testing.T) {
+	md := testonly.NewMemDev(t, 1)
+	start, length := uint(1), uint(len(md)-2)
+	j, err := OpenJournal(md, start, length)
+	if err != nil {
+		t.Fatalf("OpenJournal: %v", err)
+	}
+	if j.start != start {
+		t.Fatalf("Journal.start=%d, want %d", j.start, start)
+	}
+	if j.length != length {
+		t.Fatalf("Journal.length=%d, want %d", j.length, length)
+	}
+}
+
+func TestWriteSizeLimit(t *testing.T) {
+	storageBlocks := uint(20)
+	md := testonly.NewMemDev(t, storageBlocks)
+	start, length := uint(1), uint(storageBlocks)
+
+	j, err := OpenJournal(md, start, length)
+	if err != nil {
+		t.Fatalf("OpenJournal: %v", err)
+	}
+
+	limit := int((storageBlocks * md.BlockSize() / 3) - entryHeaderSize)
+	if err := j.Update(fill(limit, "ok...")); err != nil {
+		t.Fatalf("Update: %v, but expected write to succeed", err)
+	}
+	if err := j.Update(fill(limit+1, "BOOM")); err == nil {
+		t.Fatal("Update succeeded, but expected write fail")
+	}
+}
+
+func TestRoundTrip(t *testing.T) {
```

Comment:
Would be great to add a test that covers the case mentioned above where one of the writes perfectly fit byte-wise into the end of storage. I think that would be the first of the `doing` below, but then you don't easily get to test the wrap-around.

Reply:
hehe - I added such a test earlier :)